src/runtime-rs/crates/hypervisor/src/firecracker/fc_api.rs

```diff
@@ -0,0 +1,358 @@
+//Copyright (c) 2019-2022 Alibaba Cloud
+//Copyright (c) 2023 Nubificus Ltd
+//
+//SPDX-License-Identifier: Apache-2.0
+
+use crate::firecracker::utils::*;
+use crate::{
+    firecracker::FcInner, kernel_param::KernelParams, NetworkConfig, Param, HYPERVISOR_FIRECRACKER,
+};
+use anyhow::{anyhow, Context, Result};
+use dbs_utils::net::MacAddr;
+use hyper::{Body, Method, Request, Response};
+use hyperlocal::Uri;
+use kata_sys_util::mount;
+use nix::mount::MsFlags;
+use serde_json::json;
+use std::io::ErrorKind;
+use tokio::{fs, fs::File};
+
+const REQUEST_RETRY: u32 = 500;
+const FC_KERNEL: &str = "vmlinux";
+const FC_ROOT_FS: &str = "rootfs";
+const DRIVE_PREFIX: &str = "drive";
+const DISK_POOL_SIZE: u32 = 6;
+
+impl FcInner {
+    pub(crate) fn get_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if self.jailed {
+            self.jail_resource(src, dst)
+        } else {
+            Ok(src.to_string())
+        }
+    }
+
+    fn jail_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if src.is_empty() || dst.is_empty() {
+            return Err(anyhow!("invalid param src {} dst {}", src, dst));
+        }
+
+        let jailed_location = [
+            self.jailer_root.as_str(),
+            HYPERVISOR_FIRECRACKER,
+            &self.id,
+            ROOT,
+            dst,
+        ]
+        .join("/");
+        mount::bind_mount_unchecked(src, jailed_location.as_str(), false, MsFlags::MS_SLAVE)
+            .context("bind_mount ERROR")?;
+
+        let mut abs_path = String::from("/");
+        abs_path.push_str(dst);
+        Ok(abs_path)
+    }
+
+    // Remounting jailer root to ensure it has exec permissions, since firecracker binary will
+    // execute from there
+    pub(crate) async fn remount_jailer_with_exec(&self) -> Result<()> {
+        let dest = &[
+            &self.jailer_root,
+            "/",
+            HYPERVISOR_FIRECRACKER,
+            "/",
+            &self.id,
+            "/",
+            ROOT,
+        ]
+        .concat();
```

Comment:
`jailed_location` right above in `jail_resource()` performs a very similar operation in a slightly different way.  It might be worth it to consider making a function to perform the path assembly up to the `ROOT` element.  Also, Rust's `PathBuf` type would seem to facilitate this kind of operation.

Reply:
`PathBuf` would have been used if we had an extension or needed to push/pop a lot and we would need to mutate to string afterwards. `jailed_location` and `dest` are almost the same, but both are a single time concatenation and they already use a function for that exact purpose. I could see though ditching the `.concat()` here for the `.join()` used above.

Reply:
Okay, `PathBuf` can be a judgment call, and `join()` does seem much better than `concat()` if you insist on representing paths as `String`s.

However, the path that's anonymously constructed here (and in multiple other places in this PR) has a proper name `chroot_dir` in `jailer`'s documentation.  Also, what's apparently called `jailer_root` in this code is called `chroot_base` by `jailer`.  I do believe that it would greatly benefit readability of this code if the original jailer terminology was used consistently by the firecracker driver.

This PR assembles _a lot_ of paths, mostly in an ad-hoc manner even though some of them are related by sharing a common prefix and/or semantics.  Even where `jailer` (or firecracker) doesn't have a name for them, it would be great if this code did.

Reply:
totally agree, we tried to refactor the paths and also "tried" to use as much of the terminology fit as possible (`jailer root` != `jailer chroot dir` etc.). We've come up to the following, and we plan to clarify some more in comments and in the follow-up doc. In short it is summarized early in the `inner_hypervisor.rs` file. 

jailed case:
`jailer_path`: jailer executable
`jailer_root`: usually `/run/kata`
`vm_path`: usually `/run/kata/{vmm_exec_filename}/{sid}`
`run_dir`: usually `/run/kata/{vmm_exec_filename}/{sid}/root/run`

non-jailed case:
`jailer_path`: empty
`jailer_root`: empty
`vm_path`: usually `/run/kata/{sid}`
`run_dir`: usually `/run/kata/{sid}/run`

```
        if !self.config.jailer_path.is_empty() {
            debug!(sl!(), "[Firecracker] Running jailed");
            self.jailed = true;
            self.jailer_root = KATA_PATH.to_string();
            debug!(sl!(), "[Firecracker] jailer_root: {:?}", self.jailer_root);
            self.vm_path = [
                self.jailer_root.clone(),
                HYPERVISOR_FIRECRACKER.to_string(),
                id.to_string(),
            ]
            .join("/");
            debug!(sl!(), "[Firecracker] VM Path: {:?}", self.vm_path);
            self.run_dir = [self.vm_path.clone(), "root".to_string(), "run".to_string()].join("/");
            debug!(sl!(), "[Firecracker] Rundir: {:?}", self.run_dir);
            let _ = self.remount_jailer_with_exec().await;
        } else {
            self.vm_path = [KATA_PATH.to_string(), id.to_string()].join("/");
            debug!(sl!(), "[Firecracker] VM Path: {:?}", self.vm_path);
            self.run_dir = [self.vm_path.clone(), "run".to_string()].join("/");
            debug!(sl!(), "[Firecracker] Rundir: {:?}", self.run_dir);
        }
```

Reply:
Right, I'm wondering though, why don't you just rename `jailer_root` to `chroot_base` in the code so that it directly matches jailer terminology?

Reply:
well, no problem in doing that, however, there's no chroot_base in the kata code, only `jailer_root` (in both dragonball, CLH) so I think renaming it add another level of confusion maybe? not sure. What do the @kata-containers/rust people think? (as I'm not sure if the dragonball jailer code is in a working state... same stands for CLH I guess)

Reply:
Okay, that's a fair point, thanks.  I'd still think that the relation between firecracker/jailer and its driver in runtime-rs is rather stronger than the one between different hypervisor drivers (and folks perusing this code seem more likely to be familiar with firecracker/jailer than other hypervisor drivers) but I'm not insisting. :-)